File: apps/cli/internal/manifest/schema_bundle.go
package manifest

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"net/url"
	"sort"
	"strings"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

const schemaBundleDomain = "doppels.so/schema-bundle-sha256/v1\x00"

type schemaResource struct {
	ID       string
	Source   []byte
	Document any
}

func parseSchemaResource(source []byte) (*schemaResource, error) {
	document, err := jsonschema.UnmarshalJSON(bytes.NewReader(source))
	if err != nil {
		return nil, err
	}
	object, ok := document.(map[string]any)
	if !ok {
		return nil, errors.New("schema resource must be an object")
	}
	id, ok := object["$id"].(string)
	if !ok || id == "" {
		return nil, errors.New("schema resource must declare a non-empty $id")
	}
	parsed, err := url.Parse(id)
	if err != nil || !parsed.IsAbs() || parsed.Fragment != "" || parsed.String() != id {
		return nil, fmt.Errorf("schema resource has invalid $id %q", id)
	}
	return &schemaResource{ID: id, Source: append([]byte(nil), source...), Document: document}, nil
}

// schemaBundle resolves the complete external $ref closure for rootID. Every
// referenced resource must be supplied by the embedded, immutable schema set.
// External cycles are rejected so the bundle membership is always explicit.
func schemaBundle(rootID string, resources map[string]*schemaResource) ([]*schemaResource, error) {
	const (
		visiting = 1
		visited  = 2
	)
	state := make(map[string]int, len(resources))
	var bundle []*schemaResource
	var visit func(string) error
	visit = func(id string) error {
		resource, ok := resources[id]
		if !ok {
			return fmt.Errorf("schema bundle references unfixed external resource %q", id)
		}
		switch state[id] {
		case visiting:
			return fmt.Errorf("schema bundle contains an external $ref cycle at %q", id)
		case visited:
			return nil
		}
		state[id] = visiting
		references, err := externalSchemaReferences(resource)
		if err != nil {
			return err
		}
		for _, reference := range references {
			if err := visit(reference); err != nil {
				return err
			}
		}
		state[id] = visited
		bundle = append(bundle, resource)
		return nil
	}
	if err := visit(rootID); err != nil {
		return nil, err
	}
	sort.Slice(bundle, func(i, j int) bool { return bundle[i].ID < bundle[j].ID })
	return bundle, nil
}

func externalSchemaReferences(resource *schemaResource) ([]string, error) {
	base, err := url.Parse(resource.ID)
	if err != nil {
		return nil, err
	}
	references := map[string]struct{}{}
	var walk func(any, *url.URL) error
	walk = func(value any, currentBase *url.URL) error {
		switch typed := value.(type) {
		case map[string]any:
			nextBase := currentBase
			if rawID, present := typed["$id"]; present {
				id, ok := rawID.(string)
				if !ok || id == "" {
					return fmt.Errorf("schema %s contains a non-string or empty $id", resource.ID)
				}
				resolved, err := currentBase.Parse(id)
				if err != nil {
					return fmt.Errorf("schema %s contains invalid $id %q: %w", resource.ID, id, err)
				}
				nextBase = resolved
			}
			if rawReference, present := typed["$ref"]; present {
				reference, ok := rawReference.(string)
				if !ok || reference == "" {
					return fmt.Errorf("schema %s contains a non-string or empty $ref", resource.ID)
				}
				resolved, err := nextBase.Parse(reference)
				if err != nil {
					return fmt.Errorf("schema %s contains invalid $ref %q: %w", resource.ID, reference, err)
				}
				resolved.Fragment, resolved.RawFragment = "", ""
				dependency := resolved.String()
				if dependency != "" && dependency != resource.ID {
					references[dependency] = struct{}{}
				}
			}
			keys := make([]string, 0, len(typed))
			for key := range typed {
				keys = append(keys, key)
			}
			sort.Strings(keys)
			for _, key := range keys {
				if err := walk(typed[key], nextBase); err != nil {
					return err
				}
			}
		case []any:
			for _, item := range typed {
				if err := walk(item, currentBase); err != nil {
					return err
				}
			}
		}
		return nil
	}
	if err := walk(resource.Document, base); err != nil {
		return nil, err
	}
	result := make([]string, 0, len(references))
	for reference := range references {
		result = append(result, reference)
	}
	sort.Strings(result)
	return result, nil
}

// schemaBundleSHA256 hashes schema identities and the digest of their exact
// source bytes. Length prefixes and a versioned domain separator make the
// byte stream unambiguous and prevent reuse in another hashing namespace.
func schemaBundleSHA256(rootID string, resources map[string]*schemaResource) (string, error) {
	bundle, err := schemaBundle(rootID, resources)
	if err != nil {
		return "", err
	}
	hash := sha256.New()
	_, _ = hash.Write([]byte(schemaBundleDomain))
	writeBundleLength(hash, uint64(len(bundle)))
	for _, resource := range bundle {
		id := []byte(resource.ID)
		sourceDigest := sha256.Sum256(resource.Source)
		writeBundleLength(hash, uint64(len(id)))
		_, _ = hash.Write(id)
		writeBundleLength(hash, uint64(len(sourceDigest)))
		_, _ = hash.Write(sourceDigest[:])
	}
	return fmt.Sprintf("%x", hash.Sum(nil)), nil
}

type byteWriter interface {
	Write([]byte) (int, error)
}

func writeBundleLength(writer byteWriter, length uint64) {
	var encoded [8]byte
	binary.BigEndian.PutUint64(encoded[:], length)
	_, _ = writer.Write(encoded[:])
}

func schemaResourceIDs(resources []*schemaResource) string {
	ids := make([]string, len(resources))
	for index, resource := range resources {
		ids[index] = resource.ID
	}
	return strings.Join(ids, ",")
}

File: schemas/scripts/schema-bundle.mjs
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";
import { resolve } from "node:path";

export const BUNDLE_DOMAIN = Buffer.from("doppels.so/schema-bundle-sha256/v1\0", "utf8");

export function schemaResource(source) {
  const bytes = Buffer.isBuffer(source) ? source : Buffer.from(source);
  const document = JSON.parse(bytes.toString("utf8"));
  if (!document || typeof document !== "object" || Array.isArray(document)) {
    throw new Error("schema resource must be an object");
  }
  if (typeof document.$id !== "string" || !document.$id) {
    throw new Error("schema resource must declare a non-empty $id");
  }
  const id = new URL(document.$id);
  if (id.hash || id.href !== document.$id) {
    throw new Error(`schema resource has fragmented or non-canonical $id ${document.$id}`);
  }
  return { id: document.$id, source: bytes, document };
}

export async function loadSchemaResources(directory, filenames) {
  const resources = new Map();
  for (const filename of [...filenames].sort()) {
    const resource = schemaResource(await readFile(resolve(directory, filename)));
    if (resources.has(resource.id)) throw new Error(`duplicate schema $id ${resource.id}`);
    resources.set(resource.id, resource);
  }
  return resources;
}

export function schemaBundle(rootId, resources) {
  const state = new Map();
  const bundle = [];

  function visit(id) {
    const resource = resources.get(id);
    if (!resource) throw new Error(`schema bundle references unfixed external resource ${id}`);
    if (state.get(id) === "visiting") {
      throw new Error(`schema bundle contains an external $ref cycle at ${id}`);
    }
    if (state.get(id) === "visited") return;
    state.set(id, "visiting");
    for (const reference of externalReferences(resource)) visit(reference);
    state.set(id, "visited");
    bundle.push(resource);
  }

  visit(rootId);
  return bundle.sort((left, right) => Buffer.compare(Buffer.from(left.id), Buffer.from(right.id)));
}

function externalReferences(resource) {
  const references = new Set();

  function walk(value, base) {
    if (Array.isArray(value)) {
      for (const item of value) walk(item, base);
      return;
    }
    if (!value || typeof value !== "object") return;

    let nextBase = base;
    if (Object.hasOwn(value, "$id")) {
      if (typeof value.$id !== "string" || !value.$id) {
        throw new Error(`schema ${resource.id} contains a non-string or empty $id`);
      }
      nextBase = new URL(value.$id, base).href;
    }
    if (Object.hasOwn(value, "$ref")) {
      if (typeof value.$ref !== "string" || !value.$ref) {
        throw new Error(`schema ${resource.id} contains a non-string or empty $ref`);
      }
      const resolved = new URL(value.$ref, nextBase);
      resolved.hash = "";
      if (resolved.href !== resource.id) references.add(resolved.href);
    }
    for (const key of Object.keys(value).sort()) walk(value[key], nextBase);
  }

  walk(resource.document, resource.id);
  return [...references].sort();
}

function u64be(value) {
  const encoded = Buffer.alloc(8);
  encoded.writeBigUInt64BE(BigInt(value));
  return encoded;
}

export function schemaBundleSha256(rootId, resources) {
  const bundle = schemaBundle(rootId, resources);
  const hash = createHash("sha256");
  hash.update(BUNDLE_DOMAIN);
  hash.update(u64be(bundle.length));
  for (const resource of bundle) {
    const id = Buffer.from(resource.id, "utf8");
    const sourceDigest = createHash("sha256").update(resource.source).digest();
    hash.update(u64be(id.length));
    hash.update(id);
    hash.update(u64be(sourceDigest.length));
    hash.update(sourceDigest);
  }
  return hash.digest("hex");
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  const directory = resolve(import.meta.dirname, "..");
  const resources = await loadSchemaResources(directory, [
    "common.schema.json",
    "space.schema.json",
    "capability.schema.json",
    "recipe.schema.json"
  ]);
  for (const id of [
    "https://doppels.so/schemas/v1alpha1/capability.schema.json",
    "https://doppels.so/schemas/v1alpha1/recipe.schema.json"
  ]) {
    console.log(`${id} ${schemaBundleSha256(id, resources)}`);
  }
}
